From 5ee986d566a99a82372f6ea9f338915b91f3295c Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Sun, 1 Oct 2023 23:05:41 -0400
Subject: [PATCH 3/7] Fix rapidly pressing extension keyboard keys only sending
 half

The existing logic handles double-taps separately. It used to send a
single keypress if the key isn't a lockable modifier, but this is
obviously buggy since a double-tap is actually two presses.

Updating:
- Diff the KeyboardKey class to ensure the method names are up-to-date.
- Test to ensure rapidly pressing extension keys works correctly.

diff --git a/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali b/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
index a431ccf..4bf7589 100644
--- a/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
+++ b/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
@@ -183,6 +183,11 @@
 
     if-eq v0, v1, :cond_0
 
+    # key is not lockable (e == 0), so send the second tap (note that this
+    # method is the double tap handler, and n is the single tap handler) instead
+    # of just ignoring it
+    invoke-virtual {p0}, Lcom/realvnc/viewer/android/ui/KeyboardKey;->n()V
+
     goto :goto_0
 
     :cond_0
